lib/services/local/generic/generic153_test.go

```diff
@@ -0,0 +1,277 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package generic
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/backend/memory"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// testResource for testing the generic service.
+type testResource153 struct {
+	Metadata *headerv1.Metadata
+}
+
+func (t *testResource153) GetMetadata() *headerv1.Metadata {
+	return t.Metadata
+}
+
+var _ types.ResourceMetadata = (*testResource153)(nil)
+var _ Resource = (*resourceMetadataAdapter[*testResource153])(nil)
+
+func newTestResource153(name string) *testResource153 {
+	tr := &testResource153{
+		Metadata: &headerv1.Metadata{
+			Name: name,
+		},
+	}
+	return tr
+}
+
+// marshalResource153 marshals a generic resource.
+func marshalResource153(resource *testResource153, opts ...services.MarshalOption) ([]byte, error) {
+	return utils.FastMarshal(resource)
+}
+
+// unmarshalResource153 unmarshals a generic resource.
+func unmarshalResource153(data []byte, opts ...services.MarshalOption) (*testResource153, error) {
+	if len(data) == 0 {
+		return nil, trace.BadParameter("missing resource data")
+	}
+	cfg, err := services.CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var h types.ResourceHeader
+	if err := utils.FastUnmarshal(data, &h); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var r testResource153
+	if err := utils.FastUnmarshal(data, &r); err != nil {
+		return nil, trace.BadParameter(err.Error())
+	}
+
+	if r.Metadata == nil {
+		r.Metadata = &headerv1.Metadata{}
+	}
+
+	if cfg.ID != 0 {
+		//nolint:staticcheck // SA1019. Deprecated, but still needed.
+		r.Metadata.Id = cfg.ID
+	}
+	if cfg.Revision != "" {
+		r.Metadata.Revision = cfg.Revision
+	}
+	if !cfg.Expires.IsZero() {
+		r.Metadata.Expires = timestamppb.New(cfg.Expires)
+	}
+	return &r, nil
+}
+
+// TestGenericCRUD153 tests backend operations with the generic service.
+func TestGenericCRUD153(t *testing.T) {
+	ctx := context.Background()
+
+	memBackend, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clockwork.NewFakeClock(),
+	})
+	require.NoError(t, err)
+
+	const backendPrefix = "generic_prefix"
+
+	service, err := NewService153[*testResource153](memBackend,
```

Comment:
Maybe use a real resource, like machineid Bots, and forgo the fake?

Reply:
There are no suitable types at the moment; for example bots have no equivalent for `marshalResource153` implemented. This works, I see no reason for reworking this needlessly.